src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
                                       Task *task) {
   RAY_CHECK(worker != nullptr && task != nullptr);
   *task = worker->GetAssignedTask();
-  auto it = pinned_task_arguments_.find(task->GetTaskSpecification().TaskId());
-  RAY_CHECK(it != pinned_task_arguments_.end());
-  num_pinned_task_arguments_ -= it->second.size();
-  pinned_task_arguments_.erase(it);
+  ReleaseTaskArgs(task->GetTaskSpecification().TaskId());
   if (worker->GetAllocatedInstances() != nullptr) {
     ReleaseWorkerResources(worker);
   }
 }
 
+bool ClusterTaskManager::PinTaskArgsIfMemoryAvailable(const TaskSpecification &spec,
+                                                      bool *args_missing) {
+  std::vector<std::unique_ptr<RayObject>> args;
+  const auto &deps = spec.GetDependencyIds();
+  if (!deps.empty()) {
+    // This gets refs to the arguments stored in plasma. The refs should be
+    // deleted once we no longer need to pin the arguments.
+    if (!pin_task_arguments_(deps, &args)) {
+      *args_missing = true;
+      return false;
+    }
+    for (size_t i = 0; i < deps.size(); i++) {
+      if (args[i] == nullptr) {
+        // This can happen if the task's arguments were all local at some
+        // point, but then at least one was evicted before the task could
+        // be dispatched to a worker.
+        RAY_LOG(INFO)
+            << "Task " << spec.TaskId() << " argument " << deps[i]
+            << " was evicted before the task could be dispatched. This can happen "
+               "when there are many objects needed on this node. The task will be "
+               "scheduled once all of its dependencies are local.";
+        *args_missing = true;
+        return false;
+      }
+    }
+  }
+
+  *args_missing = false;
+  size_t task_arg_bytes = 0;
+  for (auto &arg : args) {
+    task_arg_bytes += arg->GetSize();
+  }
+  PinTaskArgs(spec, std::move(args));
```

Comment:
It's a bit confusing we have pin_task_args_ and PinTaskArgs, consider renaming one or the other.